### PR TITLE
[FF-A] ACPI TPM2 Table for FF-A Implementation:

### PR DESCRIPTION
**Pull request by Raymond-MS**
## Description

Added the TCG2 ACPI changes for FF-A enablement in order to generate the correct ACPI table with FF-A information. This includes the .inf, .asl. and .c files.

For details on how to complete these options and their meaning refer to [CONTRIBUTING.md](https://github.com/microsoft/mu/blob/HEAD/CONTRIBUTING.md).

- [x] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?
- [ ] Backport to release branch?

## How This Was Tested

Utilized the UEFI shell "acpiview" command to verify the TPM2 table is being populated properly.

## Integration Instructions

N/A

### CODE REVIEW

File: SecurityPkg/Tcg/Tcg2AcpiFfa/Tcg2AcpiFfa.c [hunk 1]
```diff
@@ -0,0 +1,750 @@
+/** @file
+  This driver implements TPM 2.0 definition block in ACPI table and
+  populates registered SMI callback functions for Tcg2 physical presence
+  and MemoryClear to handle the requests for ACPI method. It needs to be
+  used together with Tcg2 MM drivers to exchange information on registered
+  SwSmiValue and allocated NVS region address.
+
+  Caution: This module requires additional review when modified.
+  This driver will have external input - variable and ACPINvs data in SMM mode.
```
Comment (kuqin12) [inline SecurityPkg/Tcg/Tcg2AcpiFfa/Tcg2AcpiFfa.c]: This description is not true anymore.

File: SecurityPkg/Tcg/Tcg2AcpiFfa/Tcg2AcpiFfa.c [hunk 2]
```diff
@@ -0,0 +1,750 @@
...
+
+**/
+
+#include <PiDxe.h>
+
+#include <IndustryStandard/Tpm2Acpi.h>
+
+#include <Guid/TpmInstance.h>
+#include <Guid/TpmNvsMm.h>
+#include <Guid/PiSmmCommunicationRegionTable.h>
```
Comment (kuqin12) [inline SecurityPkg/Tcg/Tcg2AcpiFfa/Tcg2AcpiFfa.c]: No longer needed?

File: SecurityPkg/Tcg/Tcg2AcpiFfa/Tcg2AcpiFfa.c [hunk 3]
```diff
@@ -0,0 +1,750 @@
...
+      mTpm2AcpiTemplate.PlatformSpecificParameters[2] = 0x80; // HI Byte of Partition ID
+      mTpm2AcpiTemplate.PlatformSpecificParameters[3] = 0x02; // LO Byte of Partition ID
+      ControlArea                            = (EFI_TPM2_ACPI_CONTROL_AREA *)(UINTN)mTpm2AcpiTemplate.AddressOfControlArea;
+      ControlArea->CommandSize               = 0xF80;
+      ControlArea->ResponseSize              = 0xF80;
+      ControlArea->Command                   = PcdGet64 (PcdTpmBaseAddress) + 0x80;
+      ControlArea->Response                  = PcdGet64 (PcdTpmBaseAddress) + 0x80;
+      break;
+    default:
+      DEBUG ((DEBUG_ERROR, "TPM2 InterfaceType get error! %d\n", InterfaceType));
```
Comment (kuqin12) [inline SecurityPkg/Tcg/Tcg2AcpiFfa/Tcg2AcpiFfa.c]: I think you want to set the return status code here?

File: SecurityPkg/Tcg/Tcg2AcpiFfa/Tcg2AcpiFfa.c [hunk 4]
```diff
@@ -0,0 +1,750 @@
...
+      break;
+  }
+
+  DEBUG ((DEBUG_INFO, "Tpm2 ACPI table size %d\n", mTpm2AcpiTemplate.Header.Length));
```
Comment (kuqin12) [inline SecurityPkg/Tcg/Tcg2AcpiFfa/Tcg2AcpiFfa.c]: And then check the status code then bail the installation if failed?

File: SecurityPkg/Tcg/Tcg2AcpiFfa/Tcg2AcpiFfa.inf [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+#    TPM2.0 ACPI device object
+#     "TCG PC Client Platform Firmware Profile Specification for TPM Family 2.0 Level 00 Revision 1.03 v51"
+#       along with
+#     "Errata for PC Client Specific Platform Firmware Profile Specification Version 1.0 Revision 1.03"
+#
+#  This driver implements TPM 2.0 definition block in ACPI table and
+#  populates registered SMI callback functions for Tcg2 physical presence
+#  and MemoryClear to handle the requests for ACPI method. It needs to be
+#  used together with Tcg2 MM drivers to exchange information on registered
+#  SwSmiValue and allocated NVS region address.
```
Comment (kuqin12) [inline SecurityPkg/Tcg/Tcg2AcpiFfa/Tcg2AcpiFfa.inf]: Same here, let's update the description.

File: SecurityPkg/Tcg/Tcg2AcpiFfa/Tcg2AcpiFfa.inf [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  BaseMemoryLib
+  UefiDriverEntryPoint
+  UefiBootServicesTableLib
+  DebugLib
+  DxeServicesLib
+  TpmMeasurementLib
+  Tpm2CommandLib
+  Tcg2PhysicalPresenceLib
+  PcdLib
+  MmUnblockMemoryLib
```
Comment (kuqin12) [inline SecurityPkg/Tcg/Tcg2AcpiFfa/Tcg2AcpiFfa.inf]: Do we still need this lib? I do not think we called it anywhere?

File: SecurityPkg/Tcg/Tcg2AcpiFfa/Tcg2AcpiFfa.inf [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+[Guids]
+  gEfiTpmDeviceInstanceTpm20DtpmGuid                            ## PRODUCES           ## GUID       # TPM device identifier
+  gTpmNvsMmGuid                                                 ## CONSUMES
```
Comment (kuqin12) [inline SecurityPkg/Tcg/Tcg2AcpiFfa/Tcg2AcpiFfa.inf]: Not used?

File: SecurityPkg/Tcg/Tcg2AcpiFfa/Tcg2AcpiFfa.inf [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  gEdkiiPiSmmCommunicationRegionTableGuid                       ## CONSUMES
+
+[Protocols]
+  gEfiAcpiTableProtocolGuid                                     ## CONSUMES
+  gEfiMmCommunicationProtocolGuid                               ## CONSUMES
```
Comment (kuqin12) [inline SecurityPkg/Tcg/Tcg2AcpiFfa/Tcg2AcpiFfa.inf]: Not used?

File: SecurityPkg/Tcg/Tcg2AcpiFfa/Tcg2AcpiFfa.inf [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+[FixedPcd]
+  gEfiSecurityPkgTokenSpaceGuid.PcdSmiCommandIoPort             ## CONSUMES
```
Comment (kuqin12) [inline SecurityPkg/Tcg/Tcg2AcpiFfa/Tcg2AcpiFfa.inf]: Not used?

File: SecurityPkg/Tcg/Tcg2AcpiFfa/Tpm2Ffa.asl [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+{
+  Scope (_SB) {
+    Device(TPM0) {
+      Name (_HID, "NNNN0000")
+      Name (_CID, "MSFT0101")
+
+      //
+      // Operational region for TPM access
+      //
+      OperationRegion (TPMR, SystemMemory, 0x10000010000, 0x5000)
```
Comment (kuqin12) [inline SecurityPkg/Tcg/Tcg2AcpiFfa/Tpm2Ffa.asl]: We can change this over to use fixed PCD now, right?

File: SecurityPkg/Tcg/Tcg2AcpiFfa/Tpm2Ffa.asl [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+        {
+          QWordMemory (
+            ResourceConsumer, 
+            PosDecode, 
+            MinFixed, 
+            MaxFixed, 
+            Cacheable,
+            ReadWrite, 
+            0x0, 
+            0x10000010000,
```
Comment (kuqin12) [inline SecurityPkg/Tcg/Tcg2AcpiFfa/Tpm2Ffa.asl]: Same here?

**Comment by kuqin12** [comment]
The change seems legit. There are a few functions looks largely alike with those in the ACPI driver. When we upstream this to EDK2, I think they will want to share these functions instead of duplicating another one.

**Comment by kuqin12** [comment]
CI builds are failing, try to follow these guide to repro the failure locally: https://github.com/tianocore/tianocore.github.io/wiki/How-to-Build-With-Stuart#before-each-build-steps. There are logs published along with the pipeline builds. But it will not be efficient as the pipeline queue could be very long.

File: SecurityPkg/Tcg/Tcg2AcpiFfa/Tcg2AcpiFfa.c (hunk 3 above)
Comment (Raymond-MS) [reply SecurityPkg/Tcg/Tcg2AcpiFfa/Tcg2AcpiFfa.c]: I will change this to an if check and return EFI_UNSUPPORTED if the type isn't CRB.

File: SecurityPkg/Tcg/Tcg2AcpiFfa/Tpm2Ffa.asl [hunk 12]
```diff
@@ -0,0 +1,124 @@
+DefinitionBlock (
```
Comment (kuqin12) [inline SecurityPkg/Tcg/Tcg2AcpiFfa/Tpm2Ffa.asl:10]: copyright and file description here.

**Review by kuqin12 — APPROVED** [review]

**Pull request merged**